arm/proofs/instruction.ml

```diff
@@ -1128,6 +1208,64 @@ let arm_UZIP1 = define
           let meven:(64)word = usimd8 (\x. word_subword x (0,8): (8)word) m in
           (Rd := (word_join meven neven:(128)word)) s`;;
 
+
+let word_split_lohi = new_definition
+ `(word_split_lohi:(N tybit0)word->((N)word # (N)word)) x =
+    (word_subword x (0,dimindex(:N)):(N)word,
+     word_subword x (dimindex(:N),dimindex(:N)):(N)word)`;;
+
+(* Given x = [xlo; xhi] (LSB to MSB) and y = [ylo; yhi],
+   return [xlo;ylo;xhi;yhi]. *)
+let word_interleave2 = new_definition
+ `(word_interleave2:(N tybit0)word->(N tybit0)word->((N tybit0)tybit0)word)
+      x y =
+    let xlo,xhi = word_split_lohi x in
+    let ylo,yhi = word_split_lohi y in
+    word_join (word_join yhi xhi:(N tybit0)word)
+              (word_join ylo xlo:(N tybit0)word)`;;
+
+(* Given x = [x0;x1;x2;x3] (LSB to MSB) and y = [y0;y1;y2;y3],
+   return [x0;y0;x1;y1;x2;y2;x3;y3]. *)
+let word_interleave4 = new_definition
+ `(word_interleave4:((N tybit0)tybit0)word->((N tybit0)tybit0)word
+      ->(((N tybit0)tybit0)tybit0)word)
+      x y =
+    let xlo,xhi = word_split_lohi x in
+    let ylo,yhi = word_split_lohi y in
+    word_join (word_interleave2 xhi yhi) (word_interleave2 xlo ylo)`;;
+
+let word_interleave8 = new_definition
+ `(word_interleave8:(((N tybit0)tybit0)tybit0)word
+      ->(((N tybit0)tybit0)tybit0)word
+      ->((((N tybit0)tybit0)tybit0)tybit0)word)
+      x y =
+    let xlo,xhi = word_split_lohi x in
+    let ylo,yhi = word_split_lohi y in
+    word_join (word_interleave4 xhi yhi) (word_interleave4 xlo ylo)`;;
```

Comment:
I am fine with moving these definitions to `Library/words.ml` if you think it is necessary.